### ISSUE TITLE
Remove Room Inheritance

### ISSUE THREAD

**Issue by iampremo** [YoYoGames/GameMaker-Bugs#2923]
### Is your feature request related to a problem?

Room Inheritance has many fundamental issues with its functionality, complicates general development of features and the Room Editor, and there are far better ways of achieving the same results.

### Describe the solution you'd like

Remove the functionality of room inheritance.

**Comment by ParodyKnaveBob** [comment]
@TheNocturne @LeonardoCunhaGitHub @paxer

Methinks you misunderstood the assignment. $;^ ]

@rwkay said,
> this is your opportunity to tell us what you are actually using it for, so we can work out what needs to be done to replace it.

Two of you said your projects **rely on** / **require** it -- but didn't say how. The other said there are plenty of use cases -- but didn't say what. "What are you actually using it for?" Get your desired/needed features reinstated properly! $:^ )

**Comment by ADAbiteFIJ** [comment]
Hello everyone ! Just discovering this thread.

I really hope that room inheritance won't be removed entirely. Having inheritance between objects and room seems to be minimum to request when properly managing your code.

I wonder how so many people seem to disregard it. It is ("would be" actually because it is not working) perfect to be able to make changes in all the subsequent rooms with one action.

In my case we use it this way : 
My projet heavily focus on hours passing and every three in-game days you loop back to the morning of the first one. Each day is different and every hour NPC are moving.
- In our parent room we have 32 layers for every half-hours between 9am to 2am. It allows us to easily put everyone where we need via the room editor.
- We also have the needed layers to display every interface "item" . I know could be handling this with code but I find it so much user friendly like that.

The other use we have of this system is : 
- We have for each "scene" (room) three sub-child, one to handle each day. Because PNJ placement are not the same, many part of the environment assets must change, etc. (I could have all in one room but it's way to many layers and it's far from being ideal)

In simple terms, what I would need is : 
- Being able to delete a layer from the main parent and have it remove from all child and sub child
- Idem but with adding one layer
- When creating a child I want it to inherit every property and layers from its parents (Which is already the case and could also be achieved by copy-pasting)
**- At least, I think we all need a way to edit a group of rooms. Even if it's not "proper" inheritance.**

Thank you GameMaker team and keep up the good work !

**Comment by LeonardoCunhaGitHub** [comment]
> @TheNocturne @LeonardoCunhaGitHub @paxer
> 
> 
> 
> Methinks you misunderstood the assignment. $;^ ]
> 
> 
> 
> @rwkay said,
> 
> > this is your opportunity to tell us what you are actually using it for, so we can work out what needs to be done to replace it.
> 
> 
> 
> Two of you said your projects **rely on** / **require** it -- but didn't say how. The other said there are plenty of use cases -- but didn't say what. "What are you actually using it for?" Get your desired/needed features reinstated properly! $:^ )

Hi,

I am developing a metroidvania.
I created a "room template" for each area (more than 15 areas) and each room inherits from the template (more than 400 rooms).

The template has about 10 layers (ui, items, enemies, tiles, backgrounds...) and these and inherited by each room.

So when I create a new room I need the layers to be inherited.

When I rename a layer, add or remove I need the rooms to inherit and update to the new structure.

Each room template have code on "create event" to initialize some objects, load sounds, prefetch image groups etc... (a lot of logic here) and again the layers are involved in this code.

So with so many rooms, inheritance is a big feature here. A simple layer rename has would force me to update all rooms manually and that cost me about an hour.

The inheritance for me is fundamental.

Thank you for hearing!

**Comment by gnysek** [comment]
Template for room, and room inheritance doesn't need to be same thing, remember it.

**Comment by phablix** [comment]
> > Template for room, and room inheritance doesn't need to be same thing, remember it.
> 
> Correct.
> 
> I just wanted to clarify that...
> 
> Room template and inheritance are not the same thing.
> 
> Room template is just a room used as a model so other rooms will have the same structure (layers... code creation... backgrounds...) as a result of a duplication.
> 
> It's just a technique of implementation.
> 
> Inheritance is a gamemaker feature where children rooms inherit properties from a given parent room.
> 
> It occurs on room creation or modification.
> 
> Inheritance plays an important role here.
> 
> Room A Room B Room C Room D ...
> 
> Room A is the parent room. Also being referred as a template (as the model we want to use to have its properties replicated by its children).
> 
> Modify room A and the children will inherit the changes. Create another child room and it will inherit the properties from room A.
> 
> I will stop using the term template to avoid confusion.

I also use room inheritance for a similar use-case the problem is that sometimes updating layers don't work breaking multiple child rooms easly or just not updating them, messing with the depths etc.

I think that room inheritance is very important in every project, it saves hours of work. 
Would be great if they fix it, replace it or whatever, but something that works please...

In my case I'm waiting until they release the new room editor, hoping that they add a parent system that works.

Right now I'm scared of creating more rooms because I don't have a proper way to control the child layers without having to modify one by one (in each child room!)

So I'm focusing in other aspects of my project while I wait as I don't see myself losing time manually updating rooms.

**Comment by time-killer-games** [comment]
I haven't had the need for it.

**Comment by LeonardoCunhaGitHub** [comment]
> I haven't had the need for it.

How do you work with multiple rooms?
You don't need same layer structure across your game levels?
... or you do it all manually?
And how do you think your project structure impacts others using inheritance?

**Comment by time-killer-games** [comment]
> > I haven't had the need for it.
> 
> How do you work with multiple rooms? You don't need same layer structure across your game levels? ... or you do it all manually? And how do you think your project structure impacts others using inheritance?

Well I've been using GameMaker since GM 8.1 and GMS 1.4 days and am used to those versions more than I am GMS 2.x though i do use GMS 2.x now and have since its launch i technically didn't even know room inheritance was a thing until i saw this ticket.

**Comment by PrismaticRealms** [comment]
My two cents:  Room inheritance should not go away.  I am not familiar with the bugs that exist in the current implementation so I would say either fix them, or replace the feature with a better functioning one, but don't get rid of the concept entirely.

**Comment by Totobal5** [comment]
I think everyone here agree on the NEED of room templates that room inheritance gives but with many other problems.

**Comment by time-killer-games** [comment]
> My two cents: Room inheritance should not go away. I am not familiar with the bugs that exist in the current implementation so I would say either fix them, or replace the feature with a better functioning one, but don't get rid of the concept entirely.

I find this to be a good compromise.

**Comment by KeeVeeGames** [comment]
My current project relies on room inheritance. I am using a base room as a template and creating all levels as a child of that room. Why it won't work with just duplicating the template room or using some new same-idea functionality: as we're advancing in development we are adding new features and mechanics that need to be re-reflected in the structures of all level rooms each time. One-time duplicating a template without a "link" to that template, will still require to remade all dozens of levels made before that kind of change. The current system was mostly working with some weird quirks.
